### DISCUSSION TITLE
Opaque types/schemas?

### DISCUSSION THREAD

**Discussion by yamin-oanda** [kcl-lang/kcl#656] [Ideas]
Hi, any plans to introduce opaque types/schemas? I.e. those which may not be directly instantiated from a different module. These can be useful to enforce rules. E.g.,

```
# config.k

type Env = "test" | "prod"

opaque schema App[_env: Env]:
  name: str = "app-${_env}"

app = lambda env: Env -> App {
  App(_env = env)
}
```

And the idea is that people must use `config.app("...")` to create instances of `App`, they can't instantiate it directly with `App { ... }`.

This would prevent bypassing the generation of correct `name` values.

EDIT: alternatively, `opaque` could allow calling `App(...)` as a function which could save the boilerplate of writing the `app` lambda.

**Comment by Peefy** [comment]
OK, I see How about using the export keyword or internal keyword to label an identifier:

```python
# config.k

type Env = "test" | "prod"

schema App[_env: Env]:
  name: str = "app-${_env}"

app = lambda env: Env -> App {
  App(_env = env)
}

internal App # or export app
```

**Reply by yamin-oanda** [reply]
That might work too! Simplest way might be to not allow instantiating a schema in a different module if the schema name starts with `_`.

**Reply by Peefy** [reply]
Yes! We will consider following up with `_` Implement this feature.

**Comment by BrutalSimplicity** [comment]
Just a random watcher here, but I like the idea of being explicit about the visibility of members in public APIs rather than implicit like Python's `_` convention and Golang's capitalized first letter. I think implicit visibility can cause confusion for developers coming from other languages and could also limit the author's intent. Perhaps starting with an underscore would be appropriate in someone's use case.
